Feature preview: Bootspecs as a NixOS module with tests (#4)
DeterminateSystems/nixpkgs: doc/languages-frameworks/cuelang.section.md

# Cue (Cuelang) {#cuelang}

[Cuelang](https://cuelang.org/) is a language to:

> - describe schemas and validate backward-compatibility ;

Reviewer: ```suggestion
- describe schemas and validate backward-compatibility
```

- generate code and schemas in various formats (e.g. JSON Schema, OpenAPI)
- do configuration akin to [Dhall Lang](https://dhall-lang.org/)
- perform data validation

## Cuelang schema quick start

> Cuelang schemas are close to JSON, here is a quick cheatsheet:

Reviewer: ```suggestion
Cuelang schemas are similar to JSON, here is a quick cheatsheet:
```

- Default types includes: `null`, `string`, `bool`, `bytes`, `number`, `int`, `float`, lists as `[...T]` where `T` is a type.

> - All structures, defined by: `myStructName: { <fields> }` are **open** : they accept fields which are not specified.
> - Closed structures can be built by doing `myStructName: close({ <fields> })` : they are strict in what they accept.

Reviewer: ```suggestion
- All structures, defined by: `myStructName: { <fields> }` are **open** -- they accept fields which are not specified.
- Closed structures can be built by doing `myStructName: close({ <fields> })` -- they are strict in what they accept.
```

- `#X` are **definitions**, referenced definitions are **recursively closed**.

> - `&` operator is the unification operator, `|` is the disjunction operator.

Reviewer: Could you briefly explain what "recursively closed" means? I imagine it's similar to the opposite of "open" as described above (i.e. doesn't accept unspecified fields), but the recursive part is throwing me.

I'd also appreciate an explanation of "unification" vs "disjunction", if possible.

Author: You can have a top level closed struct containing an open struct as a field.

Recursively closed means that *all* child struct are *closed*.

I can put links, but do you think an inline explanation is good? Disjunction is quite intuitive: `A | B` is "must unify with A or must unify with B", unification is type "compatibility", `3` unify with `int`, but not with `string`, `int & string` unify with `_` (the bottom type, which has no inhabitant), `{ a: 3 } & { a: int }` unify with `{ a: 3 }`, etc.

`&` is best thought as an operator to constrain types or "inherit" structures (because constraining an open struct is just constraining the types of undeclared fields.)

Reviewer: If you can figure a good way to word it, I think an inline explanation would be good. Otherwise, I think direct links to upstream's explanations of how those operators work would be perfect. The `&` does seem very intuitive after thinking about it like "inheritance", but I'm not quite sure I grasp `|` yet :laughing: (due in no part to your explanation; it's just flying over my head at the moment lol)

Author: `a: 4 | 2`  means you can have value 4 or 2, same for: `a: int | string | List[...int]` : int or string or list of int.
Pushed a new version with links to it!

Reviewer: Oooh, that makes much more sense to me -- thanks!

Author: (Note that in Cuelang, value *ARE* types, that's why unification is used, because `3` unifies as well with `3` or `int` or `3 | int`, etc. There are some neat mathematics behind of this using a ordered lattice for compatibility relation.)

- Values **are** types, i.e. `myStruct: { a: 3 }` is a valid type definition that only allows `3` as value.

- Read <https://cuelang.org/docs/concepts/logic/> to learn more about the semantics.
- Read <https://cuelang.org/docs/references/spec/> to learn about the language specification.

## `writeCueValidator`

nixpkgs offers `pkgs.writeCueValidator` which is a helper to write validation script based on Cuelang.

Here is an example:
```

> pkgs.writeCueValidator (pkgs.writeText ''
> #Def1: {
>   field1: string
> }
> '') { document = "#Def1"; }

Reviewer: Could we more creatively indent this snippet? Otherwise it's a little difficult to grasp which is the "first parameter" and which is the "second parameter". Maybe something like this:

```suggestion
pkgs.writeCueValidator
  (pkgs.writeText ''
    #Def1: {
      field1: string
    }
  '')
  { document = "#Def1"; }
```

```

- The first parameter is the Cue schema file.
- The second paramter is an options parameter, currently, only: `document` can be passed.

`document` : match your input data against this fragment of structure or definition, e.g. you may use the same schema file but differents documents based on the data you are validating.

Another example:
```
let genericValidator = version: pkgs.writeCueValidator (pkgs.writeText ''
#Version1: {
  field1: string
}
#Version2: #Version1 & {
  field2: string
}'') { document = "#Version{toString version}" };
in
{
  validateV1 = genericValidator 1;
  validateV2 = genericValidator 2;
}
```

> The resulting is a script which takes the data you want to validate as a first argument and validate it.

Reviewer: ```suggestion
The result is a script that will validate the file you pass as the first argument against the schema you provided `writeCueValidator`.
```

Reviewer: I also think it would be valuable to have an example that uses the `validateV#` scripts. Something as simple as "here is how you use it: `"${validateV1}  path/to/document/to/validate.json"`" or something would be fine.

It can be any format that `cue vet` supports, i.e. YAML or JSON for example.

> **Known limitations** : The script will enforce **concrete** values and will not accept lossy transformations (strictness). You can add these options if you need them.

Reviewer: ```suggestion
**Known limitations**

* The script will enforce **concrete** values and will not accept lossy transformations (strictness). You can add these options if you need them.
```

I imagine there may become more known limitations as they, well, become known :P Might as well make it a list to make the introduction of those future limitations a smaller diff.

Author: These could disappear :P — they are just default choices I took because I did not want to grow this small utility too far before someone would start using it.

Reviewer: Makes sense to me!
